x/marketmap/types/path.go

```diff
@@ -0,0 +1,213 @@
+		}
+	}
+
+	return nil
+}
+
+// UniqueTickers returns all of the unique tickers across all of the paths that
+// are part of the PathsConfig. This is particularly useful for determining the
+// set of markets that are required for a given ticker as well as ensuring
+// that a given set of providers can provide the required markets.
+func (c *PathsConfig) UniqueTickers() map[Ticker]struct{} {
+	seen := make(map[Ticker]struct{})
+
+	for _, path := range c.Paths {
+		for _, ticker := range path.GetTickers() {
+			seen[ticker] = struct{}{}
+		}
+	}
+
+	return seen
+}
+
+// NewPath returns a new Path instance. A Path is a list of convertable markets
+// that will be used to convert the prices of a set of tickers to a common ticker.
+func NewPath(ops ...Operation) (Path, error) {
+	p := Path{
+		Operations: ops,
+	}
+
+	if err := p.ValidateBasic(); err != nil {
+		return Path{}, err
+	}
+
+	return p, nil
+}
+
+// Match returns true if the path matches the provided ticker.
+func (p *Path) Match(ticker string) bool {
+	if len(p.Operations) == 0 {
+		return false
+	}
+
+	first := p.Operations[0]
+	base := first.Ticker.Base
+	if first.Invert {
+		base = first.Ticker.Quote
+	}
+
+	last := p.Operations[len(p.Operations)-1]
+	quote := last.Ticker.Quote
+	if last.Invert {
+		quote = last.Ticker.Base
+	}
+
+	return ticker == fmt.Sprintf("%s/%s", base, quote)
+}
+
+// GetTickers returns the set of tickers in the path. Note that some of the tickers
+// may need to be inverted. This function does NOT return the inverted tickers.
+func (p *Path) GetTickers() []Ticker {
+	tickers := make([]Ticker, len(p.Operations))
+	for i, op := range p.Operations {
+		tickers[i] = op.Ticker
+	}
+	return tickers
+}
+
+// ShowRoute returns the route of the path in human readable format.
+func (p *Path) ShowRoute() string {
+	hops := make([]string, len(p.Operations))
+	for i, op := range p.Operations {
+		base := op.Ticker.Base
+		if op.Invert {
+			base = op.Ticker.Quote
+		}
+
+		quote := op.Ticker.Quote
+		if op.Invert {
+			quote = op.Ticker.Base
+		}
+
+		hops[i] = fmt.Sprintf("%s/%s", base, quote)
+	}
+
+	return strings.Join(hops, " -> ")
+}
+
+// ValidateBasic performs basic validation on the Path. Specifically this will check
+// that order is topologically sorted for each market. For example, if the oracle
+// receives a price for BTC/USDT and USDT/USD, the order must be BTC/USDT -> USDT/USD.
+// Alternatively, if the oracle receives a price for BTC/USDT and USD/USDT, the order
+// must be BTC/USDT -> USD/USDT (inverted == true).
+func (p *Path) ValidateBasic() error {
+	if len(p.Operations) == 0 {
+		return fmt.Errorf("path cannot be empty")
+	}
+
+	first := p.Operations[0]
+	if err := first.ValidateBasic(); err != nil {
+		return err
+	}
+
+	quote := first.Ticker.Quote
+	if first.Invert {
+		quote = first.Ticker.Base
+	}
+
+	// Ensure that the path is a directed acyclic graph.
+	seen := map[Ticker]struct{}{
+		first.Ticker: {},
+	}
+	for _, op := range p.Operations[1:] {
```

Comment:
Don't we need a size check on this?

we should see if `len(p.Operations) >= 1`

Reply:
to prevent panic nil access

Reply:
it returns an empty slice in that case, check the test case where there is only one operation.